Add docs for batched procedures (#950)
memgraph/docs: docs/reference-guide/query-modules/implement-custom-query-modules/custom-query-module-example.md

    return mgp.Record(created_vertex=vertex)

```
### Batched read procedures

> Similar to regular `read` and `write` procedures, we have `batched` `read` and `write` procedures. Batched procedures are very similar to regular procedures. The key difference is that batched procedures return results in batches, mostly to reduce memory consumption. For batched procedures, you need to define **three** functions:
> * `batching` function - similar to the main function in regular procedures
> * `initialization` function - function to initialize stream, open source file, etc...
> * `cleanup` function - function to close a stream, source file, etc...

Reviewer: ```suggestion
Similar to regular `read` and `write` procedures, Memgraph also includes `batched` `read` and `write` procedures. Batched procedures are very similar to regular procedures. The key difference is that batched procedures return results in batches, mostly to reduce memory consumption. For batched procedures, you need to define **three** functions:
* `batching` function - similar to the main function in regular procedures
* `initialization` function - function to initialize stream, open source file, etc.
* `cleanup` function - function to close a stream, source file, etc.
```

Since there are three functions, construct works as follows: 

> - `initialization` function must define to receive the same parameters in the same order as batching function including `mgp.ProcCtx` if defined as the first parameter
> - when calling procedure from the query, you need to call the `batching` function
> - Memgraph calls `initialization` before the `batching` function 
> - `batching` function needs to return an empty result at some point, which signals the end of the stream
> - `cleanup` function is called at end of the stream

Reviewer: ```suggestion
- `initialization` function must be defined in a way it receives the same parameters in the same order as batching function, including `mgp.ProcCtx` if it's defined as the first parameter
- when calling the procedure from the query, you need to call the `batching` function
- Memgraph calls `initialization` before the `batching` function 
- `batching` function needs to return an empty result at some point, which signals the end of the stream
- `cleanup` function is called at the end of the stream
```

Reviewer: if it's ok fix the same thing in the write procedures chapter

Author: Done

> There is no decorator used to register batched read procedure, but we use the `mgp` function `mgp.add_batch_read_proc('batch', 'init', 'cleanup')`

Reviewer: ```suggestion
There is no decorator used to register batched read procedure, so use the `mgp` function `mgp.add_batch_read_proc('batch', 'init', 'cleanup')`
```

def migrate(
    table_or_sql: str,
    config: mgp.Map,
) -> mgp.Record(row=mgp.Map):
    
    global mysql_dict
    cursor = mysql_dict["cursor"]
    column_names = mysql_dict["column_names"]
    rows = cursor.fetchmany(1000)
    return [mgp.Record(row=_name_row_cells(row, column_names)) for row in rows]

def cleanup_migrate():
    global mysql_dict
    mysql_dict["cursor"] = None
    mysql_dict["connection"].close()
    mysql_dict["connection"].commit()
    mysql_dict["connection"] = None
    mysql_dict["column_names"] = None
    mysql_dict = None

mgp.add_batch_read_proc(migrate, init_migrate, cleanup_migrate)

def _name_row_cells(row_cells, column_names) -> Dict[str, Any]:
    return dict(map(lambda column, value: (column, value), column_names, row_cells))
```

### Batched write procedures

Similar to batched `read` procedures, you can define batched `write` procedures. Batched procedures can return results in batches, mostly to reduce memory consumption. For batch `write` procedures like for batched `read` procedures you need to define **three** functions:
* `batching` function - similar to the main function in regular procedures

> * `initialization` function - function to initialize stream, open source file, etc...
> * `cleanup` function - function to close a stream, source file, etc...

Reviewer: ```suggestion
* `initialization` function - function to initialize stream, open source file, etc.
* `cleanup` function - function to close a stream, source file, etc.
```

Since there are three functions, construct works as follows: 
- `initialization` function must define to receive the same parameters in the same order as batching function including `mgp.ProcCtx` if defined as the first parameter
- when calling procedure from the query, you need to call the `batching` function
- Memgraph calls `initialization` before the `batching` function 
- `batching` function needs to return an empty result at some point, which signals the end of the stream
- `cleanup` function is called at end of the stream

> 

Reviewer: ```suggestion
```

> There is no decorator used to register batched read procedure, but we use the `mgp` function `mgp.add_batch_write_proc('batch', 'init', 'cleanup')`

Reviewer: ```suggestion
There is no decorator used to register batched read procedure, so use the `mgp` function `mgp.add_batch_write_proc('batch', 'init', 'cleanup')`
```

> 
> 

Reviewer: ```suggestion
```

the module in a language that throws them, use exception handlers
in `mgp_init_module` and `mgp_shutdown_module` as well.

### Batched query procedures

> Similar to batched query procedures in Python, you can add batched query procedures in C.

Reviewer: Do we expect people to reference the Python part or know anything about it? I would kind of consider each language as a separate and individual source of knowledge

Author: I think as it is similar, I would not consider it different knowledge.

Reviewer: It's not that it's different, just that somebody who is interested in query procedures in C would and first examine how to develop query procedures in Python and have knowledge enough about it that when you say: it's similar to something in the Python API, they say "aha!"

(znaci samo me zanima dal bi netko tko ce ici pisati QM u C-u znati znanje Python APIa da kad se referenciras na Python API oni znati 100% o cemu pricas... ili python developeri citaju samo o pythonu, c dev samo o cu, c++ samo o c++... ili svi citaju sve i zanimaju ih sva 3 APIja?)

Batched procedures need 3 functions, one for each of batching, initialization, and cleanup.

```c
static void batch(const struct mgp_list *args, const struct mgp_graph *graph,
                      struct mgp_result *result, struct mgp_memory *memory) {
  ...
}

static void init(const struct mgp_list *args, const struct mgp_graph *graph,
                 struct mgp_memory *memory) {
  ...
}

static void cleanup() {
  ...
}
```

> 

Reviewer: ```suggestion
```

> The `batch` function receives the list of arguments (`args`) passed in the
> query. The parameter `result` is used to fill in the resulting records of the
> procedure. Parameters `graph` and `memory` are context parameters of the
> procedure, and they are used in some parts of the provided C API. `batch`
> needs to return at some point empty `result` which signals that the `batch` procedure is done with execution and `cleanup` can be called. `init` doesn't receive `result` as it is only used for initialization. `init` function will receive same arguments which are registered and passed to `batch` function.
> Memgraph ensures to call `init` before the `batch` function, and `cleanup` at the end. The user invokes directly the `batch` function through OpenCypher.
> The passed in `memory` argument is only alive throughout the execution of
> `mgp_init_module`, so you must not allocate any global resources with it.  Consequently, you may want to reset any global state or release global resources
> in the `cleanup` function. 

Reviewer: ```suggestion
The `batch` function receives a list of arguments (`args`) passed in the
query. The parameter `result` is used to fill in the resulting records of the
procedure. Parameters `graph` and `memory` are context parameters of the
procedure, and they are used in some parts of the provided C API. 

At some point, `batch` needs to return an empty `result` to signal that the `batch` procedure is done with execution and `cleanup` can be called. `init` doesn't receive `result` as it is only used for initialization. `init` function will receive same arguments which are registered and passed to the `batch` function.

Memgraph ensures to call `init` before the `batch` function and `cleanup` at the end. The user directly invokes the `batch` function through OpenCypher.

The argument passed in `memory` is only alive throughout the execution of
`mgp_init_module`, so you must not allocate any global resources with it.  Consequently, you may want to reset any global state or release global resources
in the `cleanup` function. 
```

For more information on what exactly is possible with C API, take a look at the
`mg_procedure.h` file or the [C API reference
guide](/reference-guide/query-modules/implement-custom-query-modules/api/c-api.md).

The following line contains the `mgp_init_module` function that registers procedures
that can be invoked through Cypher. Even though the example has only one
`procedure`, you can register multiple different procedures in a single module.

> Batch procedures are invoked using the `CALL <module>.<batch_procedure> ...` syntax. The
> `<module-name>` will correspond to the name of the shared library. Since we
> compile our example into `example.so`, then the module is called `example`.
> As mentioned, Memgraph ensures to call `init` before  `<batch_procedure>` and `cleanup` once `<batch_procedure>` signals end with an empty result.

Reviewer: ```suggestion
Batch procedures are invoked using the `CALL <module>.<batch_procedure> ...` syntax. The
`<module-name>` will correspond to the name of the shared library. Since the example is complied into `example.so`, the module is called `example`.
As mentioned, Memgraph ensures to call `init` before  `<batch_procedure>` and `cleanup` once `<batch_procedure>` signals end with an empty result.
```
